### PR TITLE
feat: add ingest pipeline config

### PR DESCRIPTION
**Pull request by krispetkov**
Closes: ENG-4202

# Checklist

- [ ] `make test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/indykite/.github/blob/master/commit-message.md)
- [ ] is breaking change

## Description of change

Adding ingest pipeline config

### CODE REVIEW

File: indykite/resource_ingest_pipeline.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+			},
+			ingestPipelineOperationsTypeKey: {
+				Type:        schema.TypeList,
+				Description: "List of operations which will be allowed to be used in the ingest pipeline.",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateFunc: validation.All(
+						validation.StringInSlice(ingestPipelineAllowedOperations, false),
+					),
```
Comment (arxeiss) [inline indykite/resource_ingest_pipeline.go:91]: I think you don't need `validation.All()` when you have single validation there

**Review by arxeiss — CHANGES_REQUESTED** [review]

File: indykite/resource_ingest_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			createTimeKey:  createTimeSchema(),
+			updateTimeKey:  updateTimeSchema(),
+
+			ingestPipelineSourcesTypeKey: {
+				Type:        schema.TypeList,
+				Description: "List of sources to be used in the ingest pipeline.",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
[the remark this answers is not in the document]
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go:79]: I think not at the moment, at least for the first iteration, probably after that we can introduce some restrictions. For now it's only a `[]string`

File: indykite/resource_ingest_pipeline.go (hunk 1 above)
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go]: Fixed

File: indykite/resource_ingest_pipeline.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+				MaxItems: 6,
+			},
+			ingestPipelineAppAgentTokenTypeKey: {
+				Type:        schema.TypeString,
+				Description: "Application agent token is used to identify the application space in IndyKite APIs.",
+				Required:    true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(ingestPipelineAppAgentTokenRegex, "must be valid application agent token."),
+				),
+			},
```
[the remark this answers is not in the document]
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go]: Fixed

File: indykite/resource_ingest_pipeline.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	d *diag.Diagnostics,
+	data *schema.ResourceData,
+	_ *ClientContext,
+	builder *config.NodeRequest,
+) {
+	sources := data.Get(ingestPipelineSourcesTypeKey).([]any)
+	operations := data.Get(ingestPipelineOperationsTypeKey).([]any)
+	cfg := &configpb.IngestPipelineConfig{
+		Sources:       []string{},
+		Operations:    []configpb.IngestPipelineOperation{},
```
[the remark this answers is not in the document]
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go]: Fixed

File: indykite/resource_ingest_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		AppAgentToken: data.Get(ingestPipelineAppAgentTokenTypeKey).(string),
+	}
+
+	var cfgSources = make([]string, len(sources))
+	for i, source := range sources {
+		cfgSources[i] = source.(string)
+	}
+	cfg.Sources = cfgSources
```
[the remark this answers is not in the document]
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go]: Fixed

File: indykite/resource_ingest_pipeline_test.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+				keys["operations.#"] = Equal(strconv.Itoa(len(operations)))
+				for i, op := range operations {
+					keys[fmt.Sprintf("operations.%d", i)] = Equal(indykite.IngestPipelineOperationTypesReverse[op])
+				}
+
+				sources := data.GetConfigNode().GetIngestPipelineConfig().GetSources()
+				keys["sources.#"] = Equal(strconv.Itoa(len(sources)))
+				for i, source := range sources {
+					keys[fmt.Sprintf("sources.%d", i)] = Equal(source)
+				}
```
[the remark this answers is not in the document]
Comment (krispetkov) [reply indykite/resource_ingest_pipeline_test.go]: Fixed

File: indykite/resource_ingest_pipeline.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	_ *ClientContext,
+	builder *config.NodeRequest,
+) {
+	sources := data.Get(ingestPipelineSourcesTypeKey).([]any)
+	operations := data.Get(ingestPipelineOperationsTypeKey).([]any)
+	cfg := &configpb.IngestPipelineConfig{
+		AppAgentToken: data.Get(ingestPipelineAppAgentTokenTypeKey).(string),
+	}
+
+	cfg.Sources = rawArrayToTypedArray[string](sources)
```
Comment (arxeiss) [inline indykite/resource_ingest_pipeline.go:146]: Again, out of curiosity :laughing: You are writing so many new lines and new variables always. I'm just wondering why? Do you follow some best practice I don't know? Does that have a reason, or is it just your habit? 

I'm trying to make it as small as possible, but it also must be easy to understand. 

```diff
-	sources := data.Get(ingestPipelineSourcesTypeKey).([]any)
	operations := data.Get(ingestPipelineOperationsTypeKey).([]any)
	cfg := &configpb.IngestPipelineConfig{
		AppAgentToken: data.Get(ingestPipelineAppAgentTokenTypeKey).(string),
+		Sources: rawArrayToTypedArray[string](data.Get(ingestPipelineSourcesTypeKey).([]any))
	}
-
-	cfg.Sources = rawArrayToTypedArray[string](sources)
```

**Review by arxeiss — APPROVED** [review]

File: indykite/resource_ingest_pipeline.go (hunk 7 above)
Comment (krispetkov) [reply indykite/resource_ingest_pipeline.go]: Yeah, it's not some kind of a best practice I follow 😁 just a habit or a style I have, but sometimes it leads to not really good looking code. And in the Golang domain it's more of an anti-pattern I guess...
I don't like the long and complex one-liners, so I try to be more verbose, but sometimes it's too much 🤷 

Fixed the code with half of the proposal 😁

**Pull request merged**